Fix: Modal closes immediately after launch

kidroca opened the pull request:
### Details
Converts the existing Attachment Picker implementation to a bottom docked modal. There's a sample video in the iOS section

Notable changes:
- `AttachmentPicker`: Since this is a multi index component - to keep props interface uniform it was extracted to separate file 
- Converted native `AttachmentPicker` to a class based component, preserving existing render prop interface
- Removed logic that causes a side effect that closes the attachment options modal
   - Discussed here: https://github.com/Expensify/Expensify.cash/issues/1913#issuecomment-807851083
- Added error handling to let the user know when image picking fails for reasons other than permissions instead of leaving him guessing
- The wrapped calls to RNImage picker with promises to keep a uniform and easy to follow interface

### Fixed Issues
Fixes #1913 also fixes #2198

### Tests
<!--- 
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->
#### For web, mWeb and Desktop
Found a bug related to Safari, please use another browser to test (Chrome works): https://github.com/Expensify/Expensify.cash/issues/2109

1. Open a chat
2. Press the (+) plus button to open the input menu
3. Select "Add Attachment"
4. The browser's file picker should open and you can select a file
5. When you select a file you should see File Preview modal
6. Pressing upload should upload the file and you should see the attachment/thumb inside the chat
7. Upon the File Preview closes (previous step) focus should be returned to the compose input field

#### iOS/Android
1. Open a chat
2. Press the (+) plus button to open the input menu
3. Select "Add Attachment"
4. A second modal should open and show options for: camera, gallery, document
- Selecting "Take Photo"
  - default flow: You shoot a photo, select "Use Photo" and continue to the next step (File Preview)
  - other cases:
    - pressing "Cancel" (inside the camera) closes any modals and returns you to the "Chat".
    - not granting permission should display a helpful Alert message
    - other errors should display a general error message to let the user know something did go right. You can verify this by selecting "Take Photo" on a iOS simulator - you should see a message that taking photos on a simulator is not supported
- Selecting "Choose from Gallery"
  -  default flow: You select a photo from the gallery and continue to the next step (File Preview)
  - other cases: 
    - pressing "Cancel" (inside the gallery) closes any modals and returns you to the "Chat".
    - there's error handling that should alert the user in case of an error but I don't know a way to trigger it. I tested it by manually forcing code execution through that flow
- Selecting "Choose Document"
  -  default flow: You select a document and continue to the next step (File Preview)
  - other cases: Same as in "Choose from Gallery"
6. File Preview
- At this point you should see a preview of whatever was selected using the previous options
- Pressing "Upload" should upload the file and you should see the attachment/thumb inside the chat
- Pressing (x) close on the top right corner should close the preview and not upload anything
- Upon the File Preview closes focus should be returned to the compose input field

Regarding field input focusing
- the field is no longer focused when **any** modal closes
- the field would autofocus at the end of the full attachment flow

### QA Steps
Same as above

Note that the video sample is using the iOS simulator with the software keyboard toggled on:
https://github.com/Expensify/Expensify.cash/pull/2114#issuecomment-812449033

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
![image](https://user-images.githubusercontent.com/12156624/112678713-f18f0f80-8e73-11eb-9a0e-48bb2358ee40.png)


#### Mobile Web
![image](https://user-images.githubusercontent.com/12156624/112678901-3155f700-8e74-11eb-9160-a5e253e8572a.png)


#### Desktop
![image](https://user-images.githubusercontent.com/12156624/112679320-b5a87a00-8e74-11eb-8002-18d033b6ae7c.png)


#### iOS (Video)
https://user-images.githubusercontent.com/12156624/112827336-e0721880-9096-11eb-97d4-1daa1ea243d0.mp4


#### Android
![image](https://user-images.githubusercontent.com/12156624/112683442-3c138a80-8e7a-11eb-8bf1-982e33b62e36.png)

kidroca committed:
refactor: extract attachment prop types for reuse

kidroca committed:
docs: fix attachment native documentation

kidroca committed:
docs: attachment move example usage the renderProp

kidroca committed:
refactor: convert AttachmentPicker to class and basic rendering

kidroca committed:
refactor: AttachmentPicker add image and document pick handling

kidroca committed:
docs: add method documentation for AttachmentPicker

kidroca committed:
refactor: AttachmentPicker add error handling for the document picker too

kidroca committed:
refactor: rename func to showImagePicker to be inline with showDocumentPicker

kidroca committed:
refactor: AttachmentPicker.open should throw when the interface is not respected

Not calling the render prop with the correct interface is a developer mistake
that we can catch easily if we throw an error here

kidroca committed:
Add new icons for the Attachment Picker Menu

kidroca commented:
I have read the CLA Document and I hereby sign the CLA

kidroca commented:
recheck

kidroca committed:
refactor: ReportActionCompose add separate focus/blur methods

`setIsFocused` should be used only to set/unset the highlight styles as documented

The text input already calls setIsFocused in the event of focus/blur
Added separate methods to better control this behavior
This would also remove a direct call to `setState` inside `componentDidUpdate`

kidroca committed:
optimization: ReportActionCompose blur the input before opening the first modal

On iOS when the modal closes the keyboard will appear again as focus is restored
to the last element. This creates a choppy transition between modals
E.g. create menu -> select attachment

NikkiWines commented:
going to also add @Julesssss here as a reviewer since he's got context to for issue and proposed solution

kidroca commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -78,12 +73,9 @@ class ReportActionCompose extends React.Component {
-        ) {
-            this.setIsFocused(true);
+        // When the report ID changes bring focus to the compose field
+        if (prevProps.reportID !== this.props.reportID) {
```
Hey reviewers I think we might have a merge conflict here as @parasharrajat  work needs to change this line as well #2022 

I've changed it, to exclude modal checks as they break modal to modal transition - the attachment modal would close immediately if focus is requested on the field 

If I revert this change, I'll also need to bring  the timeout back 😭

Any suggestions?

kidroca replied:
Hmm, I've actually can revert most of my changes. The UX would be a bit odd but I guess this will be improved once autofocussing is solved and the keyboard stays collapsed

Here are some samples:
#### reverting almost everything in `ReportActionCompose` to original 

##### On Simulator
https://user-images.githubusercontent.com/12156624/112962392-022ed680-914f-11eb-9158-f53e76a1a36a.mp4

##### On a Real Device

https://user-images.githubusercontent.com/12156624/112963857-7453eb00-9150-11eb-99e3-ecc842168137.mp4

There's also a video showing how the current changes work in the pull request description

Anyway I'm willing to revert my changes to  `ReportActionCompose` so that we can merge this easier if another ticket is already addressing focusing

Julesssss replied:
The other PR should actually be changing to solve the focus issue in a less general way, so the modal checks you mention won't be a conflict for much longer.

I think this is fine for now actually.

Julesssss replied:
Hmm, maybe this is required after all for the regression I [mentioned here](https://github.com/Expensify/Expensify.cash/pull/2114#issuecomment-810095724)

kidroca replied:
Changes reverted.

Julesssss commented on src/components/AttachmentPicker/propTypes.js:
```diff
@@ -0,0 +1,25 @@
+import PropTypes from 'prop-types';
+
+const propTypes = {
+    /**
+     * A renderProp with the following interface
+     *
+     * @example
+     * <AttachmentPicker>
+     * {({openPicker}) => (
+     *     <Button
+     *         onPress={() => {
+     *             openPicker({
+     *                 onPicked: (file) => {
+     *                     // Display or upload File
+     *                 },
+     *             });
+     *         }}
+     *     />
+     * )}
+     * </AttachmentPicker>
+     * */
+    children: PropTypes.func.isRequired,
+};
```
This separate file seems unnecessary for a single PropType

kidroca replied:
I've done this because:
1. the component has platform specific implementations that need to keep the same (public) props interface. 
2. this was copy pasted to both implementations - if more props are added they will have to be copy pasted as well 
3. the example usage was copy pasted too - this means that if it changes for some reason it's possible to forget to update it on both places: e.g. ["This is the web/mWeb/desktop version since..."]( https://github.com/Expensify/Expensify.cash/pull/2114/files#diff-c5016cd4c4354cbb41b3252743d1d0024af50be56ecf42566893ad7052492b2bL104) written on the native component

kidroca replied:
@Julesssss can we resolve this one?

 My goal was to save some repetition for the documentation

The props possible to pass to a component should always match the same interface no matter the underlying platform implementation. So for platform specific components you'd need to keep them in sync if they are defined in separate files

Anyway I can define two separate prop types if you say so

Julesssss replied:
Ah okay, I missed that it was reused -- I suppose that makes sense.

Julesssss commented:
> the field is no longer focused when any modal closes

This is a regression for Web & Desktop. We need to regain focus when the attachment picker closes (not for mobile though)

Also, it looks like we're not closing the AttachmentPicker menu when on item is selected, can we do this? IMO it's odd mobile UX to see the prior modal when cancelling from the gallery (for example) -- I would expect the modal to close as the gallery/camera/picker is launched

Julesssss commented:
It tested well on Web and mobile, but mobile web (iOS) didn't work for me. I see the AttachmentModal, but tapping Add Attachment does nothing. 

I also would expect the options here to match mobile?

kidroca commented:
> This is a regression for Web & Desktop. We need to regain focus when the attachment picker closes (not for mobile though)

I've also brought this up here: https://github.com/Expensify/Expensify.cash/issues/2129#issuecomment-809554367 
Autofocus should be handled differently on mobile vs web/desktop but I don't think this can be address by the current PR
 
> Also, it looks like we're not closing the AttachmentPicker menu when on item is selected, can we do this? IMO it's odd mobile UX to see the prior modal when cancelling from the gallery (for example) -- I would expect the modal to close as the gallery/camera/picker is launched

I wasn't sure about this, I can update it to close. My logic was that if I selected gallery by mistake I can press "Cancel" and return to select something else like documents

kidroca commented:
> I also would expect the options here to match mobile?

I don't think the options would match on mWeb, it would use the web component with the hidden input - that would bring whatever file picker the browser would display to you. 

Something like this:
![image](https://user-images.githubusercontent.com/12156624/112976597-bd5e6c00-915d-11eb-919a-213f21118421.png)

It's not even possible to test it right now as this issue is also affecting mobile Chrome: #2109:
![image](https://user-images.githubusercontent.com/12156624/112977268-80df4000-915e-11eb-838d-30eb3b7c8c7c.png)

kidroca committed:
reverted changes made to ReportActionCompose.componentDidUpdate

kidroca committed:
refactor: ReportActionCompose keep focus inside the input

Closing a modal (on native) will restore focus to the last element

Current modal handling that's is forcing focus after each modal closes
would break modal to modal transition when the field was not initially
focused

kidroca commented:
Updated
- reverted changes regarding "This is a regression for Web & Desktop."
- this also closes the 2nd modal when "Cancel" is picked inside the gallery/camera

Julesssss commented:
> Autofocus should be handled differently on mobile vs web/desktop but I don't think this can be address by the current PR.

Yeah, no problem. I just mentioned it in case it simplified things, clearly not :D

Julesssss commented on src/components/AttachmentPicker/index.native.js:
```diff
@@ -28,54 +35,72 @@ const documentPickerOptions = {
+ */
+function showPermissionsAlert() {
+    Alert.alert(
+        'ExpensifyCash does not have access to your camera. To enable access, tap Settings and turn on Camera.',
```
We should use 'Expensify.cash' here. Also, I think we may be able to remove this entirely? I don't see this message on physical devices, but instead seeing this:

![permissionCamera](https://user-images.githubusercontent.com/10736861/113007418-1cc47800-916e-11eb-9088-e88359fda03b.png)

Julesssss replied:
Okay, never-mind. This is built into the library, I'll raise a new issue for this problem.

kidroca replied:
@Julesssss 
On `iOS` I do see this error if I don't grant permission:
![image](https://user-images.githubusercontent.com/12156624/113016170-e6472700-9186-11eb-92ba-82f6798eeb0d.png)

I'll update the text

Now that I've checked this is probably related: https://github.com/react-native-image-picker/react-native-image-picker#android - we have declared `android.permission.CAMERA` in the andorid manifest and so we'll need to ask for Permissions when the camera option is selected. Did this work on Android before? It might be something that was handled internally previously when `RNImagePicker.showImagePicker` was used

kidroca replied:
This is what I see for Android (no modifications to current code):

#### The app asks for permissions when the Camera option is selected
![image](https://user-images.githubusercontent.com/12156624/113026592-eac50d00-9191-11eb-8278-73a68dc9df1c.png)

#### Selecting deny shows me this message on android:
![image](https://user-images.githubusercontent.com/12156624/113026896-35df2000-9192-11eb-9246-077be82a7989.png)

---

Note: with this PR we can upgrade to v3 of this library. Maybe this is handled differently there

Julesssss commented on src/components/AttachmentPicker/index.native.js:
```diff
@@ -28,54 +35,72 @@ const documentPickerOptions = {
-            if (response.error === 'Camera permissions not granted') {
-                Alert.alert(
-                    // eslint-disable-next-line max-len
-                    'ExpensifyCash does not have access to your camera. To enable access, tap Settings and turn on Camera.',
```
Should say: 'Expensify.cash does not have access to your camera. Please enable the permission and try again.'

Julesssss replied:
Wrong place, but you get the idea

kidroca replied:
@Julesssss can you take a look here: as that message also won't fit: https://github.com/Expensify/Expensify.cash/pull/2114#issuecomment-810436951

The first parameter is Alert `title` and the second is alert `content`

kidroca replied:
I've addressed this with separate title and message:

#### Android
![image](https://user-images.githubusercontent.com/12156624/113141622-9756cb80-9232-11eb-863a-6d5ff000e753.png)

#### iOS (system theme is dark)
![image](https://user-images.githubusercontent.com/12156624/113141707-b5bcc700-9232-11eb-82b6-a462824f3d6f.png)

I'll let you decide is it OK and resolve

Julesssss replied:
Will do

Julesssss commented on src/components/AttachmentPicker/index.native.js:
```diff
@@ -28,54 +35,72 @@ const documentPickerOptions = {
+        imagePickerFunc(imagePickerOptions, (response) => {
+            if (response.error) {
+                switch (response.error) {
+                    case 'Camera permissions not granted': showPermissionsAlert(); break;
+                    default: showGeneralAlert(response.error); break;
```
Would you mind switching this to the default style that we use elsewhere? (with line breaks)

kidroca replied:
I pushed a fix, but am leaving the comment unresolved so you can verify this fix properly address your comment

Julesssss commented on src/components/AttachmentPicker/index.native.js:
```diff
@@ -94,43 +119,117 @@ function getDataForUpload(fileData) {
+    constructor(...args) {
+        super(...args);
+
+        this.state = {isVisible: false, onPicked: () => {}, result: null};
```
New lines for each field.

kidroca replied:
I pushed a fix, but am leaving the comment unresolved so you can verify this fix properly address your comment

Julesssss commented on src/components/AttachmentPicker/index.native.js:
```diff
@@ -94,43 +119,117 @@ function getDataForUpload(fileData) {
+    /**
+     * After the modal closes, if an attachment was selected delegate it to the `onPicked` callback
+     */
+    onModalHide() {
```
Functions should be named for what it does, not what triggered it. Something like `setAttachment`?

Julesssss replied:
or `setResult`

Julesssss replied:
> It's not something that you can call anytime, it needs to be called specifically when the modal is fully gone 

I think it's fine to note that in the function comment. I'm not sure that the CreateMenu usage is valid tbh. It's also the same name as a ReactNativeModal prop, which is a little confusing.

`completeAttachmentSelection` seems fine

kidroca replied:
Renamed to `completeAttachmentSelection` leaving unresolved so you can verify this fix

Julesssss commented on src/components/AttachmentPicker/index.native.js:
```diff
@@ -94,43 +119,117 @@ function getDataForUpload(fileData) {
+        }
+    }
+
+    open(onPicked) {
+        if (typeof onPicked !== 'function') {
+            throw new Error(
+                `Invalid onPicked parameter. Check the children passed to: ${AttachmentPicker.displayName}`,
+            );
+        }
+
+        this.setState({isVisible: true, onPicked, result: null});
+    }
+
+    close() {
+        this.setState({isVisible: false});
+    }
```
Could you add function docs here please

kidroca replied:
I pushed a fix, but am leaving the comment unresolved so you can verify this fix properly address your comment

kidroca committed:
docs: AttachmentPicker method documentation

kidroca committed:
refactor: AttachmentPicker update switch signature

kidroca committed:
refactor: AttachmentPicker Update camera permission message

kidroca committed:
refactor: AttachmentPicker address state and usages

kidroca committed:
fix: AttachmentPicker Handle Android permissions case

Since React Native 0.60 to open App settings use:
`Linking.openSettings()` - works on both android and ios

https://stackoverflow.com/a/58026920/4834103

kidroca commented:
Updated

Included a fix for Android camera permissions - it appears on android the `response.error` is different...
![image](https://user-images.githubusercontent.com/12156624/113029524-32996380-9195-11eb-9a22-12939301a872.png)


Still needs a designer touch as the message is used as title. 
Maybe something like:

```
Alert.alert(
  'Expensify.cash does not have access to your camera', // Title
  'To enable access, tap Settings and turn on Camera.', // Message
  [...]
)
```

NikkiWines commented on src/components/AttachmentPicker/index.native.js:
```diff
@@ -28,54 +35,77 @@ const documentPickerOptions = {
+function showGeneralAlert(message) {
+    Alert.alert(
+        'An error is preventing us to use handle the attachment',
+        message, // Todo: maybe we don't want to show this as it's probably not human friendly
```
I wouldn't include the `message` here, since - as you mentioned - it might not be user-friendly. 

I would also update the error message to be `Oops. An error occurred while selecting this attachment. Please try again.`

kidroca replied:
Thanks, the intended text actually was "An error prevented us from selecting an attachment" because this can potentially happen anywhere from initially pressing "Camera" (or the other options) so the user might not have had the chance to pick something at all 

Also the first parameter of the alert is the alert modal title, having something lengthy there might get truncated on Android like here:
![image](https://user-images.githubusercontent.com/12156624/113135685-121be880-922b-11eb-9d7a-53d2003f0d32.png)
- this is already mentioned and handled separately

I'll address this and let you decide if it's good

kidroca replied:
I've update the error to appear like so:

#### Android:
![image](https://user-images.githubusercontent.com/12156624/113142366-7642aa80-9233-11eb-9f60-8e9542aa276b.png)

#### iOS (system dark theme):
![image](https://user-images.githubusercontent.com/12156624/113142589-b99d1900-9233-11eb-943e-e451b9c5dc9d.png)

kidroca replied:
Resolving this as a new thread provided the message

NikkiWines replied:
That looks good!

NikkiWines commented on src/components/AttachmentPicker/index.native.js:
```diff
@@ -94,43 +124,135 @@ function getDataForUpload(fileData) {
+     * @throws {Error} when the `onPicked` parameter is invalid
+     */
+    open(onPicked) {
+        if (typeof onPicked !== 'function') {
```
Is there a known scenario where we could be passing a non-function parameter to `open`?

kidroca replied:
No. And there shouldn't be
 
This is an insurance that whoever is using the component is notified right away if they haven't set it up correctly similar to prop types warnings, but prop types can only declare `children` is a required function and they can't validate it'll be called correctly

If we don't have an error guard here, you'll this error will just happen later when you try to call `onPicked` after selecting an attachment it's better to shoot right away so this doesn't slip undetected

I'm leaving the comment unresolved so you can verify this addresses your concern

kidroca replied:
I'm resolving this as it seems to be the only thing blocking the PR

Julesssss replied:
I get that this might help someone, but if that is its only purpose then it doesn't seem necessary. If someone creates the component elsewhere and sees an `: TypeError: undefined is not a function`, it should be pretty obvious where the issue lies.

kidroca replied:
> I get that this might help someone, but if that is its only purpose then it doesn't seem necessary. If someone creates the component elsewhere and sees an : TypeError: undefined is not a function, it should be pretty obvious where the issue lies.

If this check is removed `undefined is not a function` would only appear after you complete the flow and select a file. 

With or without this clause you will end up with an error if the component is not setup properly, the check would just bring it to your attention sooner

I see this as something that can only help. As an external contributor it was my first time seeing a modal setup in this particular render prop configuration and it was a bit confusing for me. 

Anyway if it doesn't seem necessary to you I'll remove it next week, just let me know 😉

Julesssss replied:
I appreciate the reasoning. Including this might cause other to add similar pre-optimisations which we tend to try to avoid. It's not a huge problem, but better to remove it imo.

kidroca replied:
No problem, I'll remove it

NikkiWines commented on src/components/AttachmentPicker/propTypes.js:
```diff
@@ -0,0 +1,25 @@
+import PropTypes from 'prop-types';
```
Can we rename this file to `AttachmentPickerPropTypes.js` to match the convention we have for other files of this type?

NikkiWines reviewed (changes requested):
Some small comments

kidroca committed:
refactor: AttachmentPicker addressed error messages

kidroca committed:
refactor: AttachmentPicker rename propTypes to AttachmentPickerPropTypes

kidroca commented:
Updated
- improved error messages

Julesssss commented:
> ```js
> Alert.alert(
>   'Expensify.cash does not have access to your camera', // Title
>   'Please enable the permission and try again.', // Message
>   [...]
> )
> ```

I think that should be fine if we leave it as the message: 

Title: Camera Permission Required
Message: Expensify.cash does not have access to your camera', please enable the permission and try again.

Julesssss commented on src/components/AttachmentPicker/index.native.js:
```diff
@@ -28,54 +35,76 @@ const documentPickerOptions = {
+ */
+function showPermissionsAlert() {
+    Alert.alert(
+        'Expensify.cash does not have access to your camera.',
+        'Please enable the permission and try again.',
```
```
Title: Camera Permission Required
Message: Expensify.cash does not have access to your camera', please enable the permission and try again.
```

Julesssss commented on src/components/AttachmentPicker/index.native.js:
```diff
@@ -28,54 +35,76 @@ const documentPickerOptions = {
+ */
+function showGeneralAlert() {
+    Alert.alert(
+        'Oops. An error occurred while selecting an attachment.',
+        'Please try again.',
```
Title should be brief:

Title:  Attachment Error
Message: An error occurred while selecting an attachment, please try again

kidroca committed:
refactor: AttachmentPicker better alert messages

kidroca commented:
Updated
![image](https://user-images.githubusercontent.com/12156624/113175138-c893c380-9253-11eb-9670-6ac22bffce05.png)

kidroca commented:
@Julesssss @NikkiWines 
I have nothing more to do for this PR. Can you please take a look when you get a chance

Julesssss commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -109,14 +108,32 @@ class ReportActionCompose extends React.Component {
+    /**
+     * Un-focus from the composer text input
+     */
+    blur() {
+        if (this.textInput) {
+            this.textInput.blur();
+        }
+    }
```
This isn't being called. Is it supposed to be?

kidroca replied:
Currently no. Initially I used to clear focus from the input before opening a modal, so that the keyboard doesn't get in the way. 
Since we decided to revert this logic is no longer used, but I left it there as a compliment to `focus`

Julesssss replied:
Thanks for explaining, would you mind removing it in that case. We'd prefer to exclude an unused function as we may never need it.

kidroca replied:
Oki @Julesssss will do
Can I ask something for the future. It'll be easier to resolve such items if you say something like 
> This isn't being called. Is it supposed to be? We'd prefer to exclude an unused function as we may never need it.

So I know how to resolve this and skip some back and forth as I won't be able to address this today

Not that I'm in a hurry, but next week I'll be starting to work on other items, and I thought I was done with this one yesterday.

NikkiWines reviewed (commented):
This is looking great. Will re-review once conflicts are resolved and the two outstanding comments have been addressed. 

Just a note that I'm experiencing different behavior than what is shown in the iOS video in the OP re: focusing 

https://user-images.githubusercontent.com/16747822/113356013-7f488000-92f6-11eb-81bc-5bb2ba60a511.mov

kidroca commented:
> Just a note that I'm experiencing different behavior than what is shown in the iOS video in the OP re: focusing

Toggling the keyboard on the simulator will switch between these behaviors. Though for a real device the keyboard will always pop out (unless it has an external keyboard)

![image](https://user-images.githubusercontent.com/12156624/113402393-98603800-93ad-11eb-87c0-8fcaf2e5a362.png)

kidroca committed:
refactor: address review requested changes

kidroca commented:
I would like to use rebase to resolve the merge conflicts 
To preserve review history, I'll push a backup branch or a tag that will store the original commits
This way I think review history will be kept as the original commits would still exist

Relevant discussion on slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1617306783315300

kidroca commented:
Updated

> To preserve review history, I'll push a backup branch or a tag that will store the original commits
This way I think review history will be kept as the original commits would still exist

Sorry this didn't work. 😞 

You can still see what changes were requested in the Conversation tab, but they won't appear in "File Changes"

I've addressed all merge conflicts and tested that this still works as expected

NikkiWines reviewed (approved):
This looks good! Gonna leave it to @Julesssss for the final review + merge.

> Toggling the keyboard on the simulator will switch between these behaviors. Though for a real device the keyboard will always pop out (unless the device is connected to an external keyboard)

I would put this in your testing steps so that QA knows to do this if they're using a simulator. Otherwise they might think there's an issue with the PR.

Julesssss reviewed (changes requested):
Code is looking good, but I ran through the tests again and I'm seeing a regression in the menu options. The 'Upload Photo' option is displayed -- for every single platform.

![androidAttachment](https://user-images.githubusercontent.com/10736861/113696462-f5b7fa00-96c9-11eb-84c8-93b08fdaad6b.png)
![Simulator Screen Shot - iPhone 11 - 2021-04-06 at 11 16 00](https://user-images.githubusercontent.com/10736861/113696468-f6e92700-96c9-11eb-863a-76d7545530ef.png)
![Screenshot 2021-04-06 at 11 19 39](https://user-images.githubusercontent.com/10736861/113696490-fea8cb80-96c9-11eb-907d-b3d8b9af760f.png)

Julesssss reviewed (approved):
Whoops, forget that last review -- the option has just been renamed to 'Upload Photo'. Outside of the scope of this PR.

isagoico commented:
PR was a pass, only thing is that we were able to reproduce [this](https://github.com/Expensify/Expensify.cash/pull/2114#pullrequestreview-628806738)

![image](https://user-images.githubusercontent.com/44479856/113794652-06439100-9719-11eb-9e93-1bb52df8d240.png)

isagoico commented:
This [issue](https://github.com/Expensify/Expensify.cash/issues/2109) was reproducible in mWeb iOS only. On mWeb Android it was working as expected. 

Title: 
iOS mWeb - Attachment - Upload photo button does not work

## Expected Result
The 3 attachment options appear. 

## Actual Result
Nothing happens after clicking the upload photo button

## Action Performed
1. In a iOS device, go to Safari or Chrome
2. Navigate to staging.expensify.cash
3. Log in and navigate to a conversation
4. Click on the attachment button
5. Click on upload photo button. 

## Platform 

mWeb iOS ✔️ 

Version Number: 1.0.11-0

Logs: https://stackoverflow.com/c/expensify/questions/4856

Notes/Photos/Videos: Any additional supporting documentation

https://user-images.githubusercontent.com/44479856/113798727-bddca100-9721-11eb-9386-be1c9c8f674d.mp4

isagoico commented:
Step 7 of the Web steps is failing **only** on firefox browser. Chrome browser works as expected. 

Title: 
Web - Firefox browser - After File Preview closes focus does not return to the input field automatically

## Expected Result
After File Preview closes focus should be returned to the compose input field

## Actual Result
After File Preview closes focus does not return to the input field automatically

## Action Performed
1. Login in app https://staging.expensify.cash/signin
2. Open a chat
3. Press the (+) plus button to open the input menu
4. Select "Add Attachment"
5. The browser's file picker should open and you can select a file
6. When you select a file you should see File Preview modal
7. Pressing upload should upload the file and you should see the attachment/thumb inside the chat

## Platform 

Web (Firefox) ✔️ 

**Build:** 1.0.13-0

**Notes/Images/Videos**
Important: This is only happening on **Firefox**. Let me know if I should open a separate issue for this. On chrome, the PR was a pass. 

https://user-images.githubusercontent.com/44479856/113799871-ecf41200-9723-11eb-9e33-c8f5a2e92d53.mp4

kidroca commented:
> This issue was reproducible in mWeb iOS only. On mWeb Android it was working as expected.

It wouldn't do anything on desktop Safari too 

> Web - Firefox browser - After File Preview closes focus does not return to the input field automatically

I've went to the point before the changes introduced here - [v1.0.10-2](https://github.com/Expensify/Expensify.cash/tree/1.0.10-2) and the issue is still recreatable - it predates these changes

Julesssss commented:
Thanks @isagoico. We have these issues covered already: 

**iOS mWeb - Attachment - Upload photo button does not work** 
- We have a separate issue for Safari attachments, which currently doesn't work

**Web - Firefox browser - After File Preview closes focus does not return to the input field automatically**
- We're handling focus issues together, I'll add this not to the issue to ensure Firefox is tested with the eventual fix: https://github.com/Expensify/Expensify/issues/154180

isagoico commented:
Thanks for the clarification! Will check the PR off then 🙇

roryabraham commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -208,6 +215,10 @@ class ReportActionCompose extends React.Component {
+
+                                                    /* Keep last focus inside the input so that focus is restored
+                                                     on modal close. Otherwise breaks modal 2 modal transition */
+                                                    this.focus();
```
@kidroca I know this was a while ago, but could you potentially help me out with some reproduction steps for the issue you described here? i.e: If I remove this code how can I reproduce the modal 2 modal transition issue?

kidroca replied:
@roryabraham 
You won't be able to. This is a leftover that should be removed. It was brought up here: https://github.com/Expensify/Expensify.cash/pull/2022#issuecomment-820330081

---

After this line is removed only on iOS, and only when the input field is initially focused there would be a small quirk in the modal to modal transition, but it was always there (even with the setTimeout fix). It might not be an issue in prod. 
Check my comments and videos. There's a full explanation and a possible fix.

roryabraham replied:
Okay, so in your opinion is this safe to remove for now? (maybe with a follow-up to resolve some minor issues)?
